### PR TITLE
Async test client

### PR DESCRIPTION
**Pull request by kareemmahlees**
# PR Checklist

- [x] Have you followed the guidelines in `CONTRIBUTING.md`?
- [x] Have you got 100% test coverage on new code?
- [x] Have you updated the prose documentation?
- [x] Have you updated the reference documentation?

# Modifications 
- Created an `AsyncTestClient` at `starlite\testin\async_test_client\client.py`
- Created `BaseTransport`, `BaseTestClient` and `BaseLifeSpanHandler` to reduce code duplication 
- Created unit tests for the new client + a new test for the sync client

### CODE REVIEW

**Comment by kareemmahlees** [comment]
@Goldziher Hey there, sorry for this inconvenience but I did a mistake that led to deleting the whole PR.
I fixed the merge conflicts so you can kindly continue reviewing the changes.

File: starlite/testing/async_test_client/life_span_handler.py [hunk 1]
```diff
@@ -0,0 +1,12 @@
+from typing import TYPE_CHECKING
+
+from starlite.testing.base.life_span_handler_base import BaseLifeSpanHandler
+
+if TYPE_CHECKING:
+    from starlite.testing import AsyncTestClient
+    from starlite.types import LifeSpanReceiveMessage  # noqa: F401  # nopycln: import
+
+
+class LifeSpanHandler(BaseLifeSpanHandler):
```
Comment (Goldziher) [inline starlite/testing/async_test_client/life_span_handler.py:10]: this is redundant - you can simply create a single lifespan handler that receives a generic for the client argument, then you dont need to subclass this.

File: starlite/testing/async_test_client/transport.py [hunk 2]
```diff
@@ -0,0 +1,82 @@
...
+        self,
+        client: "AsyncTestClient",
+        raise_server_exceptions: bool = True,
+        root_path: str = "",
+    ) -> None:
+        BaseClientTransport.__init__(
+            self, client=client, raise_server_exceptions=raise_server_exceptions, root_path=root_path
+        )
+
+    async def handle_async_request(self, request: "Request") -> "Response":
```
Comment (Goldziher) [inline starlite/testing/async_test_client/transport.py:33]: this function is mostly duplicated code. What is the reason for this class?

File: starlite/testing/base/transport_base.py [hunk 3]
```diff
@@ -77,17 +77,18 @@ async def receive() -> "ReceiveMessage":
...
                     (k.decode("utf-8"), v.decode("utf-8")) for k, v in message.get("headers", [])
                 ]
                 context["response_started"] = True
             elif message["type"] == "http.response.body":
-                assert context["response_started"], 'Received "http.response.body" without "http.response.start".'
-                assert not context[
-                    "response_complete"
-                ].is_set(), 'Received "http.response.body" after response completed.'
+                if not context["response_started"]:
+                    raise AssertionError('Received "http.response.body" without "http.response.start".')
```
Comment (Goldziher) [inline starlite/testing/base/transport_base.py:89]: please revert this change ans use asserts. You can ignore the flake8 errors with a noqa comment. You are basically implementing the asserts manually here.

File: starlite/testing/sync_test_client/client.py [hunk 4]
```diff
@@ -149,27 +117,27 @@ def session(self) -> "CookieBackend":
...
-
-        Returns:
-            A contextmanager for a BlockingPortal.
-        """
-        if hasattr(self, "blocking_portal"):
-            yield self.blocking_portal
-        else:
-            with start_blocking_portal(backend=self.backend, backend_options=self.backend_options) as portal:
-                yield portal
+    # @property
```
Comment (Goldziher) [inline starlite/testing/sync_test_client/client.py:120]: why is this commented out?

File: starlite/testing/sync_test_client/life_span_handler.py [hunk 5]
```diff
@@ -0,0 +1,12 @@
+from typing import TYPE_CHECKING
+
+from starlite.testing.base.life_span_handler_base import BaseLifeSpanHandler
+
+if TYPE_CHECKING:
+    from starlite.testing import TestClient
+    from starlite.types import LifeSpanReceiveMessage  # noqa: F401  # nopycln: import
+
+
+class LifeSpanHandler(BaseLifeSpanHandler):
```
Comment (Goldziher) [inline starlite/testing/sync_test_client/life_span_handler.py:10]: see previous comment regarding this pattern - this is redundant.

File: starlite/testing/sync_test_client/transport.py [hunk 6]
```diff
@@ -0,0 +1,82 @@
...
+        self,
+        client: "TestClient",
+        raise_server_exceptions: bool = True,
+        root_path: str = "",
+    ) -> None:
+        BaseClientTransport.__init__(
+            self, client=client, raise_server_exceptions=raise_server_exceptions, root_path=root_path
+        )
+
+    def handle_request(self, request: "Request") -> "Response":
```
Comment (Goldziher) [inline starlite/testing/sync_test_client/transport.py:33]: I am also pretty sure the antire transport is redundant. 

You can simple have a handle_request method, and then a `handle_async_request` method that calls the other method and wraps it in async logic.

**Review by Goldziher — COMMENTED** [review]
nice improvement, but this is still WIP.

File: starlite/testing/sync_test_client/transport.py (hunk 6 above)
Comment (kareemmahlees) [reply starlite/testing/sync_test_client/transport.py]: @Goldziher I am not sure I completely understand this part, but if you could review the changes and see if I understood
right

File: starlite/testing/base/client_base.py [hunk 7]
```diff
@@ -50,7 +50,7 @@ def fake_asgi_connection(app: ASGIApp, cookies: Dict[str, str]) -> ASGIConnectio
     )
 
 
-class BaseTestClient:
+class BaseTestClient(Generic[T]):
```
Comment (Goldziher) [inline starlite/testing/base/client_base.py]: This change is not required, additionally its going to break existing uses.

```suggestion
class BaseTestClient:
```

File: starlite/testing/life_span_handler.py [hunk 8]
```diff
@@ -1,28 +1,29 @@
...
 if TYPE_CHECKING:
-    from starlite.testing import AsyncTestClient, TestClient
     from starlite.types import LifeSpanReceiveMessage  # noqa: F401  # nopycln: import
     from starlite.types import (
         LifeSpanSendMessage,
         LifeSpanShutdownEvent,
         LifeSpanStartupEvent,
     )
 
+T = TypeVar("T")
```
Comment (Goldziher) [inline starlite/testing/life_span_handler.py:15]: T is not bound to any type, it therefore is not typed.

```suggestion
T = TypeVar("T", bound=Union[AsyncTestClient, TestClient])
```

File: starlite/testing/sync_test_client/client.py [hunk 9]
```diff
@@ -117,28 +114,6 @@ def session(self) -> "CookieBackend":
...
-    # def session_backend(self) -> "BaseSessionBackend":
-    #     if not self._session_backend:
-    #         raise ImproperlyConfiguredException(
-    #             "Session has not been initialized for this TestClient instance. You can"
-    #             "do so by passing a configuration object to TestClient: TestClient(app=app, session_config=...)"
-    #         )
-    #     return self._session_backend
-
-    # @contextmanager
-    # def portal(self) -> Generator["BlockingPortal", None, None]:
```
Comment (Goldziher) [inline starlite/testing/sync_test_client/client.py]: why is this commented out?

File: starlite/testing/transport.py [hunk 10]
```diff
@@ -37,10 +42,10 @@ class SendReceiveContext(TypedDict):
     context: Optional[Any]
 
 
-class BaseClientTransport:
+class TestClientTransport(AsyncBaseTransport, BaseTransport, Generic[T]):
```
Comment (Goldziher) [inline starlite/testing/transport.py:43]: why do you need to extend any of the HTTPX transports here?

File: starlite/testing/transport.py [hunk 11]
```diff
@@ -1,27 +1,32 @@
...
         HTTPDisconnectEvent,
         HTTPRequestEvent,
         Message,
         Receive,
         ReceiveMessage,
         Send,
+        WebSocketScope,
     )
 
+T = TypeVar("T")
```
Comment (Goldziher) [inline starlite/testing/transport.py:26]: T is not bound

File: starlite/testing/base/client_base.py (hunk 7 above)
Comment (kareemmahlees) [reply starlite/testing/base/client_base.py]: @Goldziher but then what would be the type of the `app` argument?

File: starlite/testing/sync_test_client/client.py (hunk 9 above)
Comment (kareemmahlees) [reply starlite/testing/sync_test_client/client.py]: @Goldziher It is now in the `BaseTestClient` class

Comment (Goldziher) [reply starlite/testing/sync_test_client/client.py]: ok, then pls clean it

File: starlite/testing/base/client_base.py (hunk 7 above)
Comment (Goldziher) [reply starlite/testing/base/client_base.py]: my bad

**Comment by Goldziher** [comment]
this one is almost finished, lets bring it home @kareemmahlees

**Review by Goldziher — COMMENTED** [review]
Looks good. Please update the reference docs and usage docs - and then we can merge this.

File: starlite/testing/client/sync_client.py [hunk 12]
```diff
@@ -666,11 +616,6 @@ def test_something(self, test_client: TestClient) -> None:
             return {}
         return self.get_session_data()
 
-    @staticmethod
-    def _create_session_cookies(backend: "CookieBackend", data: Dict[str, Any]) -> Dict[str, str]:
-        encoded_data = backend.dump_data(data=data)
-        return {cookie.key: cast("str", cookie.value) for cookie in backend._create_session_cookies(encoded_data)}
-
     async def _set_session_data_async(self, data: Dict[str, Any]) -> None:
```
Comment (provinzkraut) [inline starlite/testing/client/sync_client.py:619]: These should go into the `BaseClient`. In the async client, they can simply be exposed, and in the sync client, they can be wrapped as-is. Currently, there's a re-implementation of them in the async client, which is unnecessary.

Comment (kareemmahlees) [reply starlite/testing/client/sync_client.py:619]: @provinzkraut Thank you, I noticed it while I was updating the docs and will fix it in the next commit.

Comment (kareemmahlees) [reply starlite/testing/client/sync_client.py]: @provinzkraut  Could you check the changes made [here](https://github.com/starlite-api/starlite/pull/1045/commits/6c4ed0fe2602ae5d6c9ea41affeb26ffccaa8cc7)!

File: starlite/testing/client/async_client.py [hunk 13]
```diff
@@ -461,26 +470,8 @@ async def delete(
...
     async def get_session_data_async(self) -> Dict[str, Any]:
-        return await self.session_backend.load_from_connection(
-            connection=fake_asgi_connection(
-                app=self.app,
-                cookies=dict(self.cookies),
-            ),
-        )
+        return await super().get_session_data_async()
+
+    async def set_session_data_async(self, data: Dict[str, Any]) -> None:
```
Comment (provinzkraut) [inline starlite/testing/client/async_client.py:476]:
```suggestion
    async def set_session_data(self, data: Dict[str, Any]) -> None:
```

Those should have the same name as the sync methods.

File: starlite/testing/client/base.py [hunk 14]
```diff
@@ -104,3 +118,84 @@ def portal(self) -> Generator["BlockingPortal", None, None]:
     def _create_session_cookies(backend: "CookieBackend", data: Dict[str, Any]) -> Dict[str, str]:
         encoded_data = backend.dump_data(data=data)
         return {cookie.key: cast("str", cookie.value) for cookie in backend._create_session_cookies(encoded_data)}
+
+    async def set_session_data_async(self, data: Dict[str, Any]) -> None:
```
Comment (provinzkraut) [inline starlite/testing/client/base.py:122]:
```suggestion
    async def _set_session_data(self, data: Dict[str, Any]) -> None:
```

These should be private methods

**Review by provinzkraut — CHANGES_REQUESTED** [review]

**Comment by kareemmahlees** [comment]
@Goldziher @provinzkraut Could you review?

File: docs/usage/9-testing/5-async-testing.md [hunk 15]
```diff
@@ -0,0 +1,86 @@
+# Async Testing
+
```
Comment (provinzkraut) [inline docs/usage/9-testing/5-async-testing.md:2]: The examples and such shouldn't be repeated for the two clients. You can use [tabs](https://facelessuser.github.io/pymdown-extensions/extensions/tabbed/) to display examples in a unified section of the docs instead.

File: docs/usage/9-testing/5-async-testing.md [hunk 16]
```diff
@@ -0,0 +1,86 @@
...
+If you like async, Starlite also has an async test client built on top of [httpx](https://github.com/encode/httpx). To use the async test client, similarly to the test client,you should pass to it an
+instance of Starlite as the `app` kwarg.
+
+Let's use the health check endpoint we used previously:
+
+```python title="my_app/main.py"
```
Comment (provinzkraut) [inline docs/usage/9-testing/5-async-testing.md:8]: Examples should be in a file within the `examples` directory.

File: starlite/testing/client/base.py [hunk 17]
```diff
@@ -0,0 +1,144 @@
...
+        else:
+            with start_blocking_portal(backend=self.backend, backend_options=self.backend_options) as portal:
+                yield portal
+
+    @staticmethod
+    def _create_session_cookies(backend: "CookieBackend", data: Dict[str, Any]) -> Dict[str, str]:
+        encoded_data = backend.dump_data(data=data)
+        return {cookie.key: cast("str", cookie.value) for cookie in backend._create_session_cookies(encoded_data)}
+
+    async def _set_session_data_async(self, data: Dict[str, Any]) -> None:
```
Comment (provinzkraut) [inline starlite/testing/client/base.py:122]:
```suggestion
    async def _set_session_data(self, data: Dict[str, Any]) -> None:
```

I don't mean to nitpick, but since there's no `_set_session_data_sync`, the `_async` suffix is a bit superfluous.

File: tests/logging_config/test_structlog_config.py [hunk 18]
```diff
@@ -14,7 +14,7 @@ def test_structlog_config_default(capsys: CaptureFixture) -> None:
     with create_test_client([], logging_config=StructLoggingConfig()) as client:
         assert client.app.logger
         assert isinstance(client.app.logger, BindableLogger)
-        client.app.logger.info("message", key="value")
+        client.app.logger.info("message", key="value")  # type: ignore [attr-defined]
```
Comment (provinzkraut) [inline tests/logging_config/test_structlog_config.py:17]: What changed that this isn't typed correctly anymore?

**Review by provinzkraut — CHANGES_REQUESTED** [review]

File: tests/logging_config/test_structlog_config.py (hunk 18 above)
Comment (kareemmahlees) [reply tests/logging_config/test_structlog_config.py:17]: @provinzkraut this is the error that arises:
```
tests\logging_config\test_structlog_config.py:17: error: "BindableLogger" has no attribute "info"  [attr-defined]
```
This is not something I have changed.

**Comment by Goldziher** [comment]
@all-contributors add @kareemmahlees code

**Comment by Goldziher** [comment]
well done!
